### PR TITLE
deepchem-Torch  environment

### PR DESCRIPTION
**Pull request by VIGNESHinZONE**
Fix https://github.com/deepchem/deepchem/issues/2561

- I have added `try` & `catch` based on a few sub-packages in deepchem like `metalearning`, `keras_models`, `rl` because these sub-packages are `tensorflow` dependent. 
- Didn't have a lot of problem in `trans` subpackage because the `TensorFlow` dependency was local.

These try catch strategy is inspired from the way we handle PyTorch dependencies - 
https://github.com/deepchem/deepchem/blob/cbab9678fa8aa480996bcb030f7fedcd9950ca3d/deepchem/models/__init__.py#L30-L32

cc: @peastman @rbharath

### CODE REVIEW

File: deepchem/models/__init__.py
Comment (peastman) [inline deepchem/models/__init__.py:6]: Model doesn't use TensorFlow.  It's the abstract parent class of all models.  If we can't import it, we can't import *any* model.

File: deepchem/trans/transformers.py
Comment (peastman) [inline deepchem/trans/transformers.py:1613]: We should figure out why transformers are using TensorFlow in the first place.  So far as I know, all transformers should be operating just on numpy arrays.

File: deepchem/models/__init__.py
Comment (VIGNESHinZONE) [reply deepchem/models/__init__.py]: Fixed it

**Comment by VIGNESHinZONE** [comment]
I was asked to work a building, a separate package for `deepchem[torch]`. In the process, I needed a `deepchem`, without `tensorflow` dependency so that `numpy==1.21` , `pytorch==1.9.0` & `matminer` could work properly. And hence I built on top of this PR directly.

@peastman I hope this is alright

**Comment by rbharath** [comment]
@VIGNESHinZONE Can you factor the torch changes out into a separate PR? It's harder to review two complex changes in one PR :). You can start a second PR building on the tensorflow dependency PR and rebase later

**Comment by VIGNESHinZONE** [comment]
Very sorry for making this PR so cumbersome @rbharath ,

I had chosen this because, in order to make build `deepchem[torch]`, I had to make it `Tensorflow` independent so that we could smoothly run `PyGeometric` & `matminer`.

The parts of code making it `Tensorflow` independent are just a few little try-catch blocks except in the `trans` subpackage.

**Review by rbharath — COMMENTED** [review]
This PR looks quite solid! Let's wait for the tensorflow dependency PR to get merged in, then we can rebase and then merge this PR afterwards

**Review by rbharath — APPROVED** [review]
I took a look and it looks like failing tests aren't related to this PR. I think we should now be ready to merge.

@peastman Do you have any more review comments?

**Comment by peastman** [comment]
Whenever possible, I think it's best to stick with `unittest` rather than writing PyTest specific code.  We used to use nosetests for running our tests.  When we switched to PyTest, it was a pain to go through and convert all the nosetests specific code, but everything written for unittest just kept working.  Some day we might switch yet again.  Then we'll have to convert anything that's PyTest-specific, but anything using unittest will probably still just keep working.

**Comment by rbharath** [comment]
Good point about pytest vs. unittest. One potential advantage of using the pytest mark over the unittest annotation is that it's easy to run just the pytorch tests (for example). Is there a good way to do something equivalent in unittest?

**Comment by peastman** [comment]
We would still be running them with PyTest.  I'm not sure what features it has for selecting tests.

**Comment by rbharath** [comment]
https://stackoverflow.com/questions/36456920/is-there-a-way-to-specify-which-pytest-tests-to-run-from-a-file discusses some ways pytest can select tests. Looking over the methods, I think pytest marks are likely the most robust method for demarcating a subset of tests spread across multiple files/modules (which is the case we have here for torch tests). Perhaps we can continue using marks for this purpose but whenever possible in the future default to unittest rather than pytest  infrastructure?

**Comment by ncfrey** [comment]
It looks like there may be ways to run only specific tests using `unittest` by defining test subsets:
https://stackoverflow.com/questions/1068246/python-unittest-how-to-run-only-part-of-a-test-file

**Comment by rbharath** [comment]
Looking over the options in the link @ncfrey added above, I think it's a bit gnarlier than the pytest mark approach. I think it might be cleaner to just stick with pytest marks for now. In the interest of unblock in @VIGNESHinZONE I'm going to go ahead and merge this PR, but if we come up with a cleaner way to do this with unittest let's make a swap back later

**Pull request merged**
